src/Common/IntervalKind.cpp

```diff
@@ -9,13 +9,13 @@ namespace ErrorCodes
     extern const int SYNTAX_ERROR;
 }
 
-Int32 IntervalKind::toAvgSeconds() const
+Float64 IntervalKind::toAvgSeconds() const
 {
     switch (kind)
     {
-        case IntervalKind::Nanosecond:
-        case IntervalKind::Microsecond:
-        case IntervalKind::Millisecond: return 0; /// fractional parts of seconds have 0 seconds
+        case IntervalKind::Nanosecond: return 0.000000001;
+        case IntervalKind::Microsecond: return 0.000001;
+        case IntervalKind::Millisecond: return 0.001;
         case IntervalKind::Second: return 1;
```

Comment:
Can you implement your logic without converting it to float? It will introduce floating point errors where we had accuracy before.

Reply:
The original PR had the same discussion :-)
https://github.com/ClickHouse/ClickHouse/pull/37628/files#r886792381

Reply:
The code changed in the meantime. The only Interval method which returns a Float is `toSeconds`. Its only user is the new nonNegativeDerive() SQL function which is okay with Floats.

The other Interval method `toAvgSeconds` still returns int. But even if it would return floats, its callers would not care at all because they just use the value to compare against some threshold or for logging.

The general problem is that sub-second durations cannot be expressed with integers. We would have to base the method on the smallest unit (nanoseconds). Which means (for example) a month will have 2629746000000000 nanoseconds. Quite unnatural, imho. So this will always be some sort of compromise.